independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java

```diff
@@ -588,27 +602,30 @@ public List<ClassPathElement> getElementsWithResource(String name) {
     public List<ClassPathElement> getElementsWithResource(String name, boolean localOnly) {
         ensureOpen(name);
 
+        boolean parentFirst = parentFirst(name, getClassPathResourceIndex());
+
         List<ClassPathElement> ret = new ArrayList<>();
-        if (parent instanceof QuarkusClassLoader && !localOnly) {
+
+        if (parentFirst && !localOnly && parent instanceof QuarkusClassLoader) {
             ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
         }
-        ClassPathElement[] classPathElements = getState().loadableResources.get(name);
-        if (classPathElements == null) {
-            return ret;
+
+        List<ClassPathElement> classPathElements = getClassPathResourceIndex().getClassPathElements(name);
+        ret.addAll(classPathElements);
+
+        if (!parentFirst && !localOnly && parent instanceof QuarkusClassLoader) {
+            ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
         }
-        ret.addAll(Arrays.asList(classPathElements));
+
         return ret;
```

Comment:
The behavior of this method was extremely odd as it was always returning the parent resources first. I fixed it to honor parent first in a subsequent commit (hoping it won't break anything - the rest have already been thoroughly tested on CI).